worker/provisioner/provisioner_task.go

```diff
@@ -853,31 +854,58 @@ func (task *provisionerTask) machineAvailabilityZoneDistribution(
 	// accommodating any supplied zone constraints.
 	// If the machine has a distribution group, assign based on lowest zone
 	// population of the distribution group machine.
-	var machineZone string
-	zoneMap := azMachineSort(task.availabilityZoneMachines)
+	// If more than one zone has the same number of machines, pick one of those at random.
+	zoneMachines := task.availabilityZoneMachines
 	if len(distGroupMachineIds) > 0 {
-		zoneMap = azMachineSort(task.populateDistributionGroupZoneMap(distGroupMachineIds))
+		zoneMachines = task.populateDistributionGroupZoneMap(distGroupMachineIds)
 	}
 
-	sort.Sort(zoneMap)
-	for _, zoneMachines := range zoneMap {
-		if !zoneMachines.MatchesConstraints(cons) {
-			task.logger.Debugf("machine %s does not match az %s: constraints do not match",
-				machineId, zoneMachines.ZoneName)
-			continue
-		}
-		if zoneMachines.FailedMachineIds.Contains(machineId) {
-			task.logger.Debugf("machine %s does not match az %s: excluded in failed machine ids",
-				machineId, zoneMachines.ZoneName)
-			continue
-		}
-		if zoneMachines.ExcludedMachineIds.Contains(machineId) {
-			task.logger.Debugf("machine %s does not match az %s: excluded machine id",
-				machineId, zoneMachines.ZoneName)
-			continue
+	// Make a map of zone machines keyed on count.
+	zoneMap := make(map[int][]*AvailabilityZoneMachine)
+	for _, zm := range zoneMachines {
+		machineCount := zm.MachineIds.Size()
+		zmList := append(zoneMap[machineCount], zm)
```

Comment:
Could this not just be?
`zoneMap[machineCount] = append(zoneMap[machineCount], zm)`